### PR TITLE
ticket#3959 Add "Call function" context menu item

### PR DESCRIPTION
**Pull request by atrosinenko**
Fixes [ticket 3959](https://trac.openmodelica.org/OpenModelica/ticket/3959) to some extent.

Implemented trivial "Invoke function" context menu item -- it just asks user for the text to insert inside `(...)` and sends this command to the `OMCProxy`. Sure, it introduces kind of "Modelica-injection", but at least for now, there seems to be many ways to execute anything on behalf of the user who run the OMEdit (using external functions and `Simulate model`, for example). Maybe it will be an issue when other ways would be access-controlled somehow, but now one need to just remember this entry point.

Meanwhile, should I validate a function before invocation -- after it the function becomes not saved, but updated inside the OMC.

### CODE REVIEW

**Comment by adeas31** [comment]
Thanks a lot for you contribution.

I personally think we need to improve this PR a bit. First of all I would say `Call Function` instead of `Invoke Function`. I thought of building a dialog on the fly that will list the function inputs with their respective names and units (if possible). The user inputs the values and then we simply send this to OMCProxy as you did and just open up the `OpenModelica Compiler CLI` window.

File: OMEdit/OMEditGUI/MainWindow.cpp [hunk 1]
```diff
@@ -661,17 +661,20 @@ void MainWindow::openResultFiles(QStringList fileNames)
   }
 }
 
+bool MainWindow::validateIfNeeded(LibraryTreeItem *pLibraryTreeItem)
+{
+  ModelWidget *pModelWidget = pLibraryTreeItem->getModelWidget();
+  return !pModelWidget || pModelWidget->validateText(&pLibraryTreeItem);
+}
+
```
Comment (adeas31) [inline OMEdit/OMEditGUI/MainWindow.cpp]: Doesn't make sense to have this function in MainWindow.
If its prime purpose is to get rid of 2 liner checks everywhere then I suggest moving it to `LibraryTreeItem` and updating all calls of `ModelWidget::validateText` to `LibraryTreeItem::validateText`.

Comment (atrosinenko) [reply OMEdit/OMEditGUI/MainWindow.cpp]: Reverted it to the regular call sequence used everywhere else.

**Comment by atrosinenko** [comment]
Fixed all issues except for the largest one: building a nice form on the fly.

**Comment by atrosinenko** [comment]
Added function arguments dialog.

**Comment by adeas31** [comment]
Thanks for fixes.

I fixed few things that I thought should be done in a different way. Read my commit message for details. I think we can merge this PR now.

**Comment by atrosinenko** [comment]
Thanks for your fixes too. :)

Frankly speaking, I cannot compile and run it right now (don't have repo at hands), but looks good to me and, after all, I can just send follow-up PR later if needed.
Just small comment: shouldn't copyright be 1998-2018 instead of 1998-2014?

**Comment by adeas31** [comment]
Yeah you are right. Actually our latest copyright text says "1998-CurrentYear". I will update the copyright text in the separate commit as I need to update it for other files as well.

**Comment by atrosinenko** [comment]
> Doesn't make sense to have this function in MainWindow.
> If its prime purpose is to get rid of 2 liner checks everywhere then I suggest moving it to  LibraryTreeItem and updating all calls of ModelWidget::validateText to LibraryTreeItem::validateText.

I have once thought, maybe these lines are really worth of extracting to some function with a meaningful name. Then I downloaded [PMD](https://pmd.github.io/) and used its copy-paste detector module on OMEdit. It seems, there are quite large parts of duplicated code (such as `OMSimulatorHighlighter::highlightMultiLine` and `HTMLHighlighter::highlightMultiLine` -- almost 2x50 lines of textually identical code) -- are these left as such by design or could they be deduplicated by someone (maybe me, if I would find time for it) without breaking third-party code of your licensees? It can probably help to make the codebase more contributor-friendly and maintainable. Maybe there are some guidelines for those who wants to clean-up the code?

**Comment by adeas31** [comment]
I agree with you that there are some parts with duplicated code but of course we should try to avoid such things and should fix the existed such code. The problem with `OMSimulatorHighlighter::highlightMultiLine` and `HTMLHighlighter::highlightMultiLine` is that they both are xml based editors and has similar highlighting implementation. Perhaps the `CompositeModelHighlighter::highlightMultiLine` is also same. I think we could introduce a base class for xml editors and implement the `highlightMultiLine` for them and if any of the child classes want different implementation they can override it.

What do you think?

(Maybe you can open a new issue about it and we follow the discussion there. This PR is already merged.)

**Comment by atrosinenko** [comment]
> What do you think?

This looks to me as a very natural OOP way of solving this issue.

> (Maybe you can open a new issue about it and we follow the discussion there. This PR is already merged.)

Sure, another issue or PR should be opened. Maybe I will investigate it further on weekend. Unfortunately, I cannot open issue in *this* repository, just on Trac (I realize that this is intentionally and most of the time preferably with such a multi-repo setup).

**Comment by adeas31** [comment]
Yeah the issues on this repository are disabled since we are still using Trac. So a Trac ticket would be fine.

**Comment by atrosinenko** [comment]
I will probably investigate it after the release, since it introduces no valuable features but can introduce last-minute bugs. :)

**Pull request merged**